stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+            // TODO: might be better to store these transactions on the side to prevent having to query the stacker db for every signer (only do on initilaization of a new signer for example and then listen for stacker db updates after that)
+            let old_transactions = self.get_signer_transactions(stacks_client, current_reward_cycle).map_err(|e| {
+                warn!("Signer #{}: Failed to get old transactions: {e:?}. Potentially overwriting our existing transactions", self.signer_id);
+            }).unwrap_or_default();
+            // Check if we have an existing vote transaction for the same round and reward cycle
+            for transaction in old_transactions.iter() {
+                let origin_address = transaction.origin_address();
+                if &origin_address != stacks_client.get_signer_address() {
+                    continue;
+                }
+                let TransactionPayload::ContractCall(payload) = &transaction.payload else {
+                    error!("BUG: Signer #{}: Received an unrecognized transaction ({}) in an already filtered list: {transaction:?}", self.signer_id, transaction.txid());
+                    continue;
+                };
+                if payload.function_name == VOTE_FUNCTION_NAME.into() {
+                    let Some((_signer_index, point, round, _reward_cycle)) =
+                        Self::parse_function_args(&payload.function_args)
+                    else {
+                        error!("BUG: Signer #{}: Received an unrecognized transaction ({}) in an already filtered list: {transaction:?}", self.signer_id, transaction.txid());
+                        continue;
+                    };
+                    if Some(point) == self.coordinator.aggregate_public_key
+                        && round == self.coordinator.current_dkg_id
+                    {
+                        debug!("Signer #{}: Not triggering a DKG round. Already have a pending vote transaction for aggregate public key {point:?} for round {round}...", self.signer_id);
+                        return Ok(());
+                    }
+                } else {
+                    error!("BUG: Signer #{}: Received an unrecognized transaction ({}) in an already filtered list: {transaction:?}", self.signer_id, transaction.txid());
+                    continue;
+                }
+            }
+            if stacks_client
+                .get_vote_for_aggregate_public_key(
+                    self.coordinator.current_dkg_id,
+                    self.reward_cycle,
+                    *stacks_client.get_signer_address(),
+                )?
+                .is_some()
+            {
+                // TODO Check if the vote failed and we need to retrigger the DKG round not just if we have already voted...
+                // TODO need logic to trigger another DKG round if a certain amount of time passes and we still have no confirmed DKG vote
```

Comment:
Are there tickets for these issues?

Reply:
There are now :P

https://github.com/stacks-network/stacks-core/issues/4418

https://github.com/stacks-network/stacks-core/issues/4419